# Analysis of polyploid genotypes generated by Sujan

## Location of files
### Raw VCF's (genotype calls, no read counts)
* Sujan's directory with VCF's
  * `/global/projectb/scratch/sujan/Pvirgatum_polyploid/`
* My directory
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs`
### Files to help process VCFs
* BED Files
  * CDS BED
    * `/global/cscratch1/sd/grabowsp/sg_ploidy/genic_positions/sg_v5_CDS.bed`
  * Genes BED
    * `/global/cscratch1/sd/grabowsp/sg_ploidy/genic_positions/sg_v5_genes.bed`
* Library names by Ploidy Call
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/tetraploid_lib_names_May2020.txt`
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/octoploid_lib_names_May2020.txt`
* Name of all "Good" Libaries
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/samp901_lib_names.txt`
### Filtered VCF Directories
* CDS VCFs
  `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs`
* CDS VCFs with 901 "good" samples
  `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps`


## Generate CDS VCFs for each chromosome
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs
sbatch generate_Chr01K_N_CDS_vcfs.sh

gunzip -kc Chr01K.polyploid.CDS.vcf.gz | head 

sbatch generate_Chr02to05_CDS_vcfs.sh
sbatch generate_Chr06to09_CDS_vcfs.sh
```
* Moved to `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs`


## CDS VCF for included samples - 'allsamp'
```
module load python/3.7-anaconda-2019.07
source activate gen_bioinformatics

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps

sbatch generate_Chr01K_01N_CDS_allsamp_vcf.sh
sbatch generate_Chr02_05_CDS_allsamp_vcf.sh
sbatch generate_Chr06_09_CDS_allsamp_vcf.sh
```
### Tests using Chr01K
#### Split vcfs into 100k-line files
##### Chr01K Files with different levels of missing data
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps

gunzip -kc Chr01K.polyploid.CDS.allsamps.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.allsamps.vcf_

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps/few_missing

gunzip -kc Chr01K.polyploid.CDS.allsamps.few_miss.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.allsamps.few_miss.vcf_

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps/half_missing

gunzip -kc Chr01K.polyploid.CDS.allsamps.half_miss.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.allsamps.half_miss.vcf_

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps/high_missing

gunzip -kc Chr01K.polyploid.CDS.allsamps.high_miss.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.allsamps.high_miss.vcf_
``` 
### Generate VCF header to be used for `allsamps` VCFs
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps
head -5 Chr01K.polyploid.CDS.allsamps.vcf_00 | tail -1 > \
CDS.allsamps.vcf.header.txt
```
### Filter CDS VCF by missing data
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/all_samps

sbatch generate_Chr01_CDS_MissFilt.sh
```

## Generate CDS VCF for natural/geographic samples
* maximum 20% missing data
* Samples sets (on Cori):
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/geo826_lib_names.txt`
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/geo_expand_839_lib_names.txt`
  * Description and code here:
    * `~/sg_ha_effort/samp_set_explanations.md`
### Make Natural Sample VCFs
#### No 8X cultivars - `geosamps`
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

sbatch generate_Chr01K_01N_CDS_geosamps_vcf.sh
sbatch generate_Chr02_Chr05_CDS_geosamps_vcf.sh
sbatch generate_Chr06_Chr09_CDS_geosamps_vcf.sh
```
#### Include 8X cultivars - `expandgeosamps`
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/expand_geo_samps

sbatch generate_Chr01K_01N_CDS_expandgeosamps_vcf.sh
sbatch generate_Chr02_Chr05_CDS_expandgeosamps_vcf.sh
sbatch generate_Chr06_Chr09_CDS_expandgeosamps_vcf.sh
```
### Split Chromosome VCFs into 100K line subfiles
#### Without 8X Cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

gunzip -kc Chr01K.polyploid.CDS.geosamps.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.geosamps.vcf_

gunzip -kc Chr01N.polyploid.CDS.geosamps.vcf.gz | \
split -l 100000 -d - Chr01N.polyploid.CDS.geosamps.vcf_

for CN in {02..09};
do
gunzip -kc Chr$CN'K.polyploid.CDS.geosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'K.polyploid.CDS.geosamps.vcf_';
done

for CN in {02..09};
do
gunzip -kc Chr$CN'N.polyploid.CDS.geosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'N.polyploid.CDS.geosamps.vcf_';
done
```
#### With 8X Cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/expand_geo_samps

gunzip -kc Chr01K.polyploid.CDS.expandgeosamps.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.expandgeosamps.vcf_

gunzip -kc Chr01N.polyploid.CDS.expandgeosamps.vcf.gz | \
split -l 100000 -d - Chr01N.polyploid.CDS.expandgeosamps.vcf_

for CN in {02..09};
do
gunzip -kc Chr$CN'K.polyploid.CDS.expandgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'K.polyploid.CDS.expandgeosamps.vcf_';
done

for CN in {02..09};
do
gunzip -kc Chr$CN'N.polyploid.CDS.expandgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'N.polyploid.CDS.expandgeosamps.vcf_';
done

```
### Generate VCFs headers
#### Without 8X Cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps
head -5 Chr01K.polyploid.CDS.geosamps.vcf_00 | tail -1 > \
CDS.geosamps.vcf.header.txt
```
#### With 8X Cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/expand_geo_samps
head -5 Chr01K.polyploid.CDS.expandgeosamps.vcf_00 | tail -1 > \
CDS.expandgeosamps.vcf.header.txt
```

## Generate 'genlight' Objects for Natural Samples
### Generate Chromosome genlight objects
#### Without 8X cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

sbatch gen_geo_genlight.sh
```
#### With 8X cultivars
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/expand_geo_samps

sbatch gen_expandgeo_genlight.sh
```
### Generate genome-wide, subsampled genlight object
#### Without 8X cultivars
* on Cori
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps/Combo.595K.polyploid.CDS.geosamps.genlight.rds`
* on HA
  * `/home/t4c1/WORK/grabowsk/data/switchgrass/polyploid_genos/genlight_objs/Combo.595K.polyploid.CDS.geosamps.genlight.rds`
##### Generate object
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps

DATA_DIR=/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps/

FILE_SUB=geosamps.genlight.rds

OUT_NAME=Combo.595K.polyploid.CDS.geosamps.genlight.rds

PER_SUBSAMP=0.07

Rscript /global/homes/g/grabowsp/tools/sg_ha_effort/polyploid_genos/adegenet_analysis/subsample_genlight.r \
$DATA_DIR '*'$FILE_SUB $OUT_NAME $PER_SUBSAMP
```

## "Upland" and "Lowland" Samples
* based on the 'geosamps' PCA results
  * so these do not include the 8X cultivars
* Cutoffs
  * upland = PC1 < -25
  * lowland = PC1 > 0
### Generate library lists
```
module load python/3.7-anaconda-2019.07
source activate r_adegenet_env

library(adegenet)

res_file <- '/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/geo_samps/Combo.595K.polyploid.CDS.geosamps.genlight.PCAresults.rds'

pca_res <- readRDS(res_file)

pca_mat <- pca_res$scores

up_samps <- rownames(pca_mat)[which(pca_mat[,1] < -25)]
low_samps <- rownames(pca_mat)[which(pca_mat[,1] > 0)]

# 18 of 826 samples are in neither group

up_out_file <- '/global/cscratch1/sd/grabowsp/sg_ploidy/upland_libs_June2020.txt'
write.table(up_samps, file = up_out_file, quote = F, sep = '\t', row.names = F, 
  col.names = F)

low_out_file <- '/global/cscratch1/sd/grabowsp/sg_ploidy/lowland_libs_June2020.txt'
write.table(low_samps, file = low_out_file, quote = F, sep = '\t', 
  row.names = F, col.names = F)
```
### Generate CDS VCFs for Upland and Lowland samples
#### Upland
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps

sbatch generate_Chr01K_01N_CDS_upgeosamps_vcf.sh
sbatch generate_Chr02_Chr05_CDS_upgeosamps_vcf.sh
sbatch generate_Chr06_Chr09_CDS_upgeosamps_vcf.sh
```
#### Lowland
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps

sbatch generate_Chr01K_01N_CDS_lowgeosamps_vcf.sh
sbatch generate_Chr02_Chr05_CDS_lowgeosamps_vcf.sh
sbatch generate_Chr06_Chr09_CDS_lowgeosamps_vcf.sh
```
### Split Chromosome VCFs into 100K line subfiles
#### Uplands
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps

for CN in {01..09};
do
gunzip -kc Chr$CN'K.polyploid.CDS.upgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'K.polyploid.CDS.upgeosamps.vcf_';
done

for CN in {01..09};
do
gunzip -kc Chr$CN'N.polyploid.CDS.upgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'N.polyploid.CDS.upgeosamps.vcf_';
done
```
#### Lowlands
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps

for CN in {01..09};
do
gunzip -kc Chr$CN'K.polyploid.CDS.lowgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'K.polyploid.CDS.lowgeosamps.vcf_';
done

for CN in {01..09};
do
gunzip -kc Chr$CN'N.polyploid.CDS.lowgeosamps.vcf.gz' | \
split -l 100000 -d - Chr$CN'N.polyploid.CDS.lowgeosamps.vcf_';
done
```
### Generate VCFs headers
#### Upland
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps
head -5 Chr01K.polyploid.CDS.upgeosamps.vcf_00 | tail -1 > \
CDS.upgeosamps.vcf.header.txt
```
#### Lowland
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps
head -5 Chr01K.polyploid.CDS.lowgeosamps.vcf_00 | tail -1 > \
CDS.lowgeosamps.vcf.header.txt
```
### Generate 'genlight' objects
#### Chromosome genlight objects
#### Upland
* use 0.006 as MAF
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps

sbatch gen_upgeo_genlight.sh
```
### Lowland
* use 0.003 as MAF
* need to make the script
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps

sbatch gen_lowgeo_genlight.sh
```
### Generate genome-wide, subsampled genlight object
#### Upland
* Goal is ~200k SNPs
  * keeps roughly the same SNP/sample ratio as with the geo samp results
* Subsample 6% of the SNPs
* on Cori:
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps/Combo.sub.polyploid.CDS.upgeosamps.genlight.rds`
* on HA:
  * `/home/t4c1/WORK/grabowsk/data/switchgrass/polyploid_genos/genlight_objs/Combo.sub.polyploid.CDS.upgeosamps.genlight.rds`
##### Generate Object
```
module load python/3.7-anaconda-2019.07
source activate r_adegenet_env

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps

DATA_DIR=/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps/

FILE_SUB=upgeosamps.genlight.rds

OUT_NAME=Combo.sub.polyploid.CDS.upgeosamps.genlight.rds

PER_SUBSAMP=0.06

Rscript /global/homes/g/grabowsp/tools/sg_ha_effort/polyploid_genos/adegenet_analysis/subsample_genlight.r \
$DATA_DIR '*'$FILE_SUB $OUT_NAME $PER_SUBSAMP
```
#### Lowland
* Goal is ~390000 SNPs
* Subsample 6.4%
* on Cori:
  * `/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps/Combo.sub.polyploid.CDS.lowgeosamps.genlight.rds`
* on HA:
  * `/home/t4c1/WORK/grabowsk/data/switchgrass/polyploid_genos/genlight_objs/Combo.sub.polyploid.CDS.lowgeosamps.genlight.rds`
##### Generate object
```
module load python/3.7-anaconda-2019.07
source activate r_adegenet_env

cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps

DATA_DIR=/global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/low_geo_samps/

FILE_SUB=lowgeosamps.genlight.rds

OUT_NAME=Combo.sub.polyploid.CDS.lowgeosamps.genlight.rds

PER_SUBSAMP=0.064

Rscript /global/homes/g/grabowsp/tools/sg_ha_effort/polyploid_genos/adegenet_analysis/subsample_genlight.r \
$DATA_DIR '*'$FILE_SUB $OUT_NAME $PER_SUBSAMP
```

## Generate CDS VCFs for Upland subgroups for looking at r^2
### Generate Files
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps/up_subgroups

sbatch generate_Chr01K_01N_CDS_up_subgroups_vcf.sh

#sbatch generate_Chr01K_01N_CDS_upgeosamps_vcf.sh
#sbatch generate_Chr02_Chr05_CDS_upgeosamps_vcf.sh
#sbatch generate_Chr06_Chr09_CDS_upgeosamps_vcf.sh
```
### Split Chromosome VCFs into 100K line subfiles
#### Start with Chr01K
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps/up_subgroups

# group tet_1
gunzip -kc Chr01K.polyploid.CDS.up_tet_1.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.up_tet_1.vcf_

gunzip -kc Chr01K.polyploid.CDS.up_tet_2.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.up_tet_2.vcf_

gunzip -kc Chr01K.polyploid.CDS.up_oct_1.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.up_oct_1.vcf_

gunzip -kc Chr01K.polyploid.CDS.up_oct_2.vcf.gz | \
split -l 100000 -d - Chr01K.polyploid.CDS.up_oct_2.vcf_
```
### Make header files
```
cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps/up_subgroups

head -5 Chr01K.polyploid.CDS.up_oct_1.vcf_00 | tail -1 > \
CDS.up_oct_1.vcf.header.txt

head -5 Chr01K.polyploid.CDS.up_oct_2.vcf_00 | tail -1 > \
CDS.up_oct_2.vcf.header.txt

head -5 Chr01K.polyploid.CDS.up_tet_1.vcf_00 | tail -1 > \
CDS.up_tet_1.vcf.header.txt

head -5 Chr01K.polyploid.CDS.up_tet_2.vcf_00 | tail -1 > \
CDS.up_tet_2.vcf.header.txt


cd /global/cscratch1/sd/grabowsp/sg_ploidy/polyploid_vcfs/CDS_vcfs/up_geo_samps
head -5 Chr01K.polyploid.CDS.upgeosamps.vcf_00 | tail -1 > \
CDS.upgeosamps.vcf.header.txt
```
